[TG-9173] Add build option to enable profiling

owen-mc-diffblue opened the pull request:
Once I have this working I will add some documentation for it somewhere

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add option to enable profiling

For profiling, -pg must be used for the compilation and the linking
steps. (-g must be used as well, so it should be a debug or
relwithdebinfo build)

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 167f5bf9d182275943e0d270a799b05b6be1e947).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123930600
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Document profiling option

Add instructions on building cprover binaries so they can be profiled
by a tool like gprof.

owen-mc-diffblue commented:
I have added documentation, including instructions for using gprof to profile where the time is spent in a particular run of a cprover binary.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a8c78239419db0dad5f7494b340f0ec6531a2620).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124722982

antlechner commented on doc/architectural/compilation-and-development.md:
```diff
@@ -270,3 +270,28 @@ branch, e.g. to run it on lines that have been changed from `develop`:
+\section compilation-and-development-section-time-profiling Time profiling
+
+To do time profiling with a tool like `gprof`, the flags `-g` and `-pg` must be
+used which compiling, and `-pg` must be used when linking. If you are building
```
:pick: Typo: `which compiling` -> `when compiling`.

antlechner commented on doc/architectural/compilation-and-development.md:
```diff
@@ -270,3 +270,28 @@ branch, e.g. to run it on lines that have been changed from `develop`:
+
+\section compilation-and-development-section-time-profiling Time profiling
+
+To do time profiling with a tool like `gprof`, the flags `-g` and `-pg` must be
```
:bulb: It would be good to provide a very short explanation of these flags, e.g. "the flags `-g` (build with debug symbols) and `-pg` (enable profiling information) must be used..."
:grey_question: Just wondering, what is the difference between using `-g` for both compiling and linking, and using it for compiling only?

owen-mc-diffblue replied:
I assume that `-g` doesn't have any effect when linking, whereas `-pg` has an effect in both compiling and linking

antlechner commented on CMakeLists.txt:
```diff
@@ -107,6 +107,18 @@ if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR
+        "${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang" OR
+        "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU"
+        )
```
:pick: Not sure if this is supposed to be at the beginning of the line (matching other similar blocks in this file).

antlechner commented on doc/architectural/compilation-and-development.md:
```diff
@@ -270,3 +270,28 @@ branch, e.g. to run it on lines that have been changed from `develop`:
+
+To do time profiling with a tool like `gprof`, the flags `-g` and `-pg` must be
+used which compiling, and `-pg` must be used when linking. If you are building
+with cmake you can just add "-Denable_profiling=1" to your cmake invocation, and
```
:bulb: I observed on my machine that if I build with cmake twice, first without this option, then with this option (without making any changes to files in between), the second build takes a long time even though I'm using ccache. This might be worth mentioning here.
:grey_question: Is the `=1` necessary?

owen-mc-diffblue replied:
I just checked and cmake prints an error without the `=1`

antlechner commented on doc/architectural/compilation-and-development.md:
```diff
@@ -270,3 +270,28 @@ branch, e.g. to run it on lines that have been changed from `develop`:
+working directory of the binary at the end of execution. In most instances this
+will be the same as the working directory at the beginning of execution. It is
+also possible to choose the output location using an environment variable -
+search online for more details.
```
:bulb: A link would be more welcoming than telling people to just "google it". :slightly_smiling_face:

owen-mc-diffblue replied:
I've google it myself and put the answer in

antlechner commented on doc/architectural/compilation-and-development.md:
```diff
@@ -270,3 +270,28 @@ branch, e.g. to run it on lines that have been changed from `develop`:
+run `sudo apt install binutils`.
+
+Run `gprof <path-to-binary> <path-to-gmon.out>` and redirect the output to a
+file. This will take a while to run.
```
:bulb: In case you want to be more specific than "a while", I ran jbmc without options on a very simple Java main method (one assignment to an int, one assert on the int) and this step took 12min for me. Of course this will depend on the machine this is run on but it would be good to document a general order of magnitude of how long this could take.

antlechner commented on doc/architectural/compilation-and-development.md:
```diff
@@ -270,3 +270,28 @@ branch, e.g. to run it on lines that have been changed from `develop`:
+Run `gprof <path-to-binary> <path-to-gmon.out>` and redirect the output to a
+file. This will take a while to run.
+
+The output file will now be a large text file. There are two sections: the
```
:bulb: The official names of the two sections are "Flat profile" and "Call graph". It would be good to mention them here as their names appear as titles, i.e. are searchable, in the giant text file.
:bulb: Apparently other kinds of sections can be generated as well if `gprof` is run with certain options. Two are mentioned on https://ftp.gnu.org/old-gnu/Manuals/gprof-2.9.1/html_chapter/gprof_5.html - this or a similar tutorial might be good to link to here for people looking for more detailed info? (Again, they could google it but a link would be friendlier.)

antlechner reviewed (approved):
Works for me using a cmake build. :+1: The documentation is helpful too, combined with a few online searches it was enough to make everything work. My comments are suggestions for how to improve it, but none of them are blocking.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0c720545287fe9d1ba98307035eb3d24b62f2377).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125245051
